crates/polars-time/src/quarter_end.rs

```diff
@@ -0,0 +1,84 @@
+use arrow::legacy::time_zone::Tz;
+use chrono::NaiveDateTime;
+use polars_core::prelude::*;
+use polars_core::utils::arrow::temporal_conversions::{
+    timestamp_ms_to_datetime, timestamp_ns_to_datetime, timestamp_us_to_datetime, MILLISECONDS,
+    SECONDS_IN_DAY,
+};
+
+use crate::quarter_start::roll_backward;
+use crate::windows::duration::Duration;
+
+// roll forward to the last day of the quarter
+fn roll_forward(
+    t: i64,
+    time_zone: Option<&Tz>,
+    timestamp_to_datetime: fn(i64) -> NaiveDateTime,
+    datetime_to_timestamp: fn(NaiveDateTime) -> i64,
+    offset_fn: fn(&Duration, i64, Option<&Tz>) -> PolarsResult<i64>,
+) -> PolarsResult<i64> {
+    let t = roll_backward(t, time_zone, timestamp_to_datetime, datetime_to_timestamp)?;
+    let t = offset_fn(&Duration::parse("3mo"), t, time_zone)?;
+    offset_fn(&Duration::parse("-1d"), t, time_zone)
+}
+
+pub trait PolarsQuarterEnd {
+    fn quarter_end(&self, time_zone: Option<&Tz>) -> PolarsResult<Self>
+    where
+        Self: Sized;
+}
+
+impl PolarsQuarterEnd for DatetimeChunked {
+    fn quarter_end(&self, time_zone: Option<&Tz>) -> PolarsResult<Self> {
```

Comment:
This function is identical down to the character to `month_end`. Can't we reuse some of the code? If we're going to have month/quarter/year end functions, and they involve computing a date offset and performing the offset, then they should all be reusing the same machinery except for the offset calculation.

Reply:
Yes, ofc
It's POC for the function and once We decided that we will implement the in polars (and not move to polars-ext, or straight up decided not to implement) i will perform refactoring